File: 189-rotate-array/solution.ts
/**
 Do not return anything, modify nums in-place instead.
 */
export function rotate(nums: number[], k: number): void {
  k = k % nums.length;
  reverseArrInplace(nums, 0, nums.length - k);
  reverseArrInplace(nums, nums.length - k, nums.length);
  reverseArrInplace(nums, 0, nums.length);
}

function reverseArrInplace<T>(arr: T[], start: number, end: number): void {
  while (start < end) {
    const startNum = arr[start] as T;
    arr[start] = arr[end - 1] as T;
    arr[end - 1] = startNum;
    ++start;
    --end;
  }
}

File: 189-rotate-array/solution.py
from typing import List


class Solution:
  def rotate(self, nums: List[int], k: int) -> None:
    """
    Do not return anything, modify nums in-place instead.
    """
    k = k % len(nums)
    Solution.__reverse_list_inplace(nums, 0, len(nums) - k)
    Solution.__reverse_list_inplace(nums, len(nums) - k, len(nums))
    Solution.__reverse_list_inplace(nums, 0, len(nums))

  @staticmethod
  def __reverse_list_inplace(li: List, start: int, end: int) -> None:
    while start < end:
      start_num = li[start]
      li[start] = li[end - 1]
      li[end - 1] = start_num
      start += 1
      end -= 1
